### PR TITLE
Allow mission descriptions to refer to the effects of the mission.

### PR DESCRIPTION
**Pull request by Petethegoat**
```SUMMARY: Interface "Allow mission descriptions to refer to the effects of the mission."```

#### Purpose of change
Make written mission descriptions more resilient to becoming outdated if they want to refer to eg. the rewards provided by the mission.

#### Describe the solution
`talk_effect_fun_t` now has a simple and specific structure to store likely mission rewards, and it fills this during json loading. `mission_type` then grabs this and stores it for use by the mission UI later, using simple string replacement.
Currently it's limited to looking for the `"u_buy_item"` effect, and only when the cost is 0 (unspecified defaults to 0).

The tokens in descriptions look like this: `"<reward_count:FMCNote>"`
They'll look for a `"u_buy_item"` effect with a matching item id, and replace the whole token with the count for that item.
This does work correctly for multiple tokens in a single description. So you can reward several things and refer to all their counts correctly.

#### Describe alternatives you've considered
This is very specific for things that are probably rewards for the mission only- but it's kept it relatively simple rather than trying to guess future use-cases ahead of time, so refactoring it to include other parameters shouldn't be too painful, if need be.

#### Additional context
Created for #33325.

### CODE REVIEW

**Comment by Petethegoat** [comment]
To be clear, it's the stuff in mission_util.cpp that makes me question if this is something usable. Maybe that can be implemented more generically somehow, but I don't have any particularly bright ideas regarding that.

**Comment by Petethegoat** [comment]
@ralreegorganon suggested a much less boneheaded way of getting the effect data that I've now implemented, which makes this feel a lot less hacky.

Right now I'm still putting the data into a very specific structure that's probably only going to be useful for my use case.
I'm not sure if I should try to generalize it without really knowing what other people will need, or if I should just keep it simple so someone who knows what they want can easily refactor it in the future.

Leaning towards the latter, lazier option, but I'm open to suggestions.

File: src/mission_ui.cpp [hunk 1]
```diff
@@ -123,10 +124,21 @@ void game::list_missions()
             y += fold_and_print( w_missions, point( 31, y ), getmaxx( w_missions ) - 33, col,
                                  miss->name() + for_npc );
 
+            auto format_tokenized_description = []( const std::string description,
+            std::vector<std::pair<int, std::string>> rewards ) {
+                std::string formatted_description = description;
+                for( size_t i = 0; i < rewards.size(); i++ ) {
+                    std::regex token("<reward_count:" + rewards[i].second + ">");
```
Comment (BevapDin) [inline src/mission_ui.cpp]: You don't need a regex for simple string replacement, we have `string_replace` for this.

File: src/mission.h [hunk 2]
```diff
@@ -371,6 +374,7 @@ class mission
         int get_id() const;
         const std::string &get_item_id() const;
         character_id get_npc_id() const;
+        std::vector<std::pair<int, std::string>> get_likely_rewards();
```
Comment (BevapDin) [inline src/mission.h]:
```suggestion
        const std::vector<std::pair<int, std::string>> &get_likely_rewards() const;
```

Same for the function in the other class.

**Comment by BevapDin** [comment]
There needs to be at least once instance that actually uses the new code, so we can see it works as expected. And documentation in some file in the "doc" folder would be nice (when it's finished).

File: src/mission_ui.cpp (hunk 1 above)
Comment (Petethegoat) [reply src/mission_ui.cpp]: I was utilizing it at one point, but refactored to this. I'll update to use `string_replace` instead.

**Comment by Petethegoat** [comment]
@BevapDin Kevin accidentally merged my Free Merchant descriptions PR that was supposed to wait for this, so there's already descriptions using this in master.

Will work on docs next, and then I think I'm happy with this, assuming the laziness regarding generalizing this is acceptable :slightly_smiling_face:

**Comment by Petethegoat** [comment]
Docs added and PR description updated.

**Comment by ZhilkinSerg** [comment]
Some conflicts creeped in. Can you resolve them please?

**Comment by Petethegoat** [comment]
Will do shortly.

**Comment by Petethegoat** [comment]
@ZhilkinSerg done, thanks for the heads up. :)

File: src/mission_ui.cpp [hunk 3]
```diff
@@ -123,10 +123,21 @@ void game::list_missions()
             y += fold_and_print( w_missions, point( 31, y ), getmaxx( w_missions ) - 33, col,
                                  miss->name() + for_npc );
 
+            auto format_tokenized_description = []( const std::string description,
+            const std::vector<std::pair<int, std::string>> rewards ) {
+                std::string formatted_description = description;
+                for( size_t i = 0; i < rewards.size(); i++ ) {
```
Comment (ZhilkinSerg) [inline src/mission_ui.cpp]:
```
/home/travis/build/CleverRaven/Cataclysm-DDA/src/mission_ui.cpp:129:17: error: use range-based for loop instead [modernize-loop-convert,-warnings-as-errors]

                for( size_t i = 0; i < rewards.size(); i++ ) {

                ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                   (const auto & reward : rewards)
```

**Pull request merged**